platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/reconnect/ReconnectNodeRemover.java

```diff
@@ -195,187 +119,50 @@ public void setPathInformation(final long newFirstLeafPath, final long newLastLe
      * Register the receipt of a new leaf node. If the leaf node is in the position that was formally occupied by
-     * a leaf node then this method will ensure that the old leaf node is properly deleted. If the leaf node is
-     * in the position that was formally occupied by an internal node then this method will ensure all leaves
-     * in the subtree are properly deleted.
+     * a leaf node then this method will ensure that the old leaf node is properly deleted.
      *
      * @param path
      * 		the path to the node
      * @param newKey
      * 		the key of the new leaf node
      */
     public void newLeafNode(final long path, final K newKey) {
-        abortAndLogIfInterrupted(
-                () -> workQueue.put(new ReceivedNode<>(path, true, newKey)),
-                "reconnect virtual map node removal thread interrupted");
+        final VirtualLeafRecord<K, ?> oldRecord = oldRecords.findLeafRecord(path, false);
+        if ((oldRecord != null) && !newKey.equals(oldRecord.getKey())) {
+            leavesToDelete.add(oldRecord);
+        }
+
+        if ((path == newLastLeafPath) && (newLastLeafPath < oldLastLeafPath)) {
+            for (long p = newLastLeafPath + 1; p <= oldLastLeafPath; p++) {
+                final VirtualLeafRecord<K, ?> oldExtraLeafRecord = oldRecords.findLeafRecord(p, false);
+                assert oldExtraLeafRecord != null || p < oldFirstLeafPath;
+                if (oldExtraLeafRecord != null) {
+                    leavesToDelete.add(oldExtraLeafRecord);
+                }
+            }
+        }
     }
 
     /**
      * Return a stream of keys that require deletion.
      *
-     * @param requiredPath
-     * 		the minimum path that must be processed before the stream is returned
      * @return a stream of keys to be deleted. Only the key and path in these records
      * 		are populated, all other data is uninitialized.
      */
-    public Stream<VirtualLeafRecord<K, V>> getRecordsToDelete(final long requiredPath) {
-        while (handledPath.get() < requiredPath && !exceptionEncountered.get() && !closed.get()) {
-            tryToSleep(Duration.ofMillis(1));
-        }
-
-        if (exceptionEncountered.get()) {
-            throw new RuntimeException("VirtualMap reconnect node removal thread has crashed");
-        }
-
-        if (handledPath.get() < requiredPath) {
-            throw new RuntimeException("VirtualMap reconnect node removal couldn't process all paths");
-        }
-
-        workQueue.pause();
-        final Stream<VirtualLeafRecord<K, V>> stream = keysToBeRemoved.entrySet().stream()
-                .map((final Map.Entry<K, Long> entry) ->
-                        new VirtualLeafRecord<>(entry.getValue(), entry.getKey(), null));
-
-        // We can't just clear the map, as doing so will disrupt the stream constructed above.
-        keysToBeRemoved = new HashMap<>();
-
-        workQueue.resume();
-
+    public Stream<VirtualLeafRecord<K, V>> getRecordsToDelete() {
```

Comment:
Usually getters are idempotent. Please rename it or, at least, mention in the javadoc that it's not idempotent.

Reply:
I added a comment to this method and also to `leavesToDelete` field in the same class.